src/Streams.Core/Streams.fs

```diff
@@ -4,20 +4,18 @@ open System.Collections.Generic
 open System.Threading
 
 /// Provides on-demand iteration 
-type Iterator = {
+type (* internal *) Iterator = 
     /// Function for on-demand processing
-    TryAdvance : unit -> bool 
+    abstract TryAdvance : unit -> bool 
     /// Cleanup function
-    Dispose : unit -> unit 
-}
+    abstract Dispose : unit -> unit 
 
 /// Provides functions for iteration
-type Iterable = {
+type (* internal *) Iterable = 
```

Comment:
Yep, I think you have a valid point! 
This https://github.com/nessos/Streams/blob/master/src/Streams.Core/Streams.fs#L140-L143 needs to be inside the contructor of Iterator